_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,364 @@
+It happens when the thread holds a monitor lock or has a native call in the stack:
+
+[source, java]
+----
+Object monitor = new Object();
+//...
+public void aMethodThatPinTheCarrierThread() throws Exception {
+  synchronized(monitor) {
+    Thread.sleep(1000); // The virtual thread cannot be unmounted because it holds a lock, 
+                        // so the carrier thread is blocked.
+  }
+}
+----
+
+In this case, the carrier thread is blocked, so the OS thread is blocked:
+
+image::pinning.png[Pinning of the carrier thread,400,float="right",align="center"]
+
+Unfortunately, as of today, lots of Java libraries are pinning the carrier thread. 
+The Quarkus team and Red Hat, in general, have patched many libraries (such as Narayana (the transaction manager of Quarkus) or Hibernate ORM) to avoid pinning. 
+However, when you use a library, be careful. 
+It will take time until all the code gets reworked in a more virtual-thread-friendly way.
+
+### 2. Monopolization
+
+As for the reactive model, if the virtual thread executes intensive and long computation, it monopolizes that carrier.
+The virtual thread scheduler is not preemptive.
+So it cannot interrupt a running thread. 
+It needs to wait for an I/O or the completion of the computation.
+Until then, this carrier thread cannot execute other virtual threads:
+
+image::monopolization.png[Monopolization of the carrier thread,400,float="right",align="center"]
+
+Using a dedicated platform thread pool might be wiser when executing long computations. 
+
+### 3. Carrier thread pool elasticity
+
+When there is pinning or monopolization, the JVM may create new carrier threads (as illustrated on the previous picture) to avoid having too many unscheduled virtual threads. 
+
+These creations are creating platform/OS threads. 
+So, it's expensive and uses memory.
+You especially need to pay attention to the second point. 
+You may hit the memory limit if you run on low resources and your code is not very virtual-thread-friendly, meaning that you should always check for pinning, monopolization, and memory usage. 
+If you don't, in a container with memory constraints, the application can be killed.
+
+### 4. Object pooling
+
+For years, threads were scarce resources.
+It was recommended to pool them and reuse them. 
+This good practice has encouraged the use of thread locals as an object-pooling mechanism.
+Like https://github.com/FasterXML/jackson-core/issues/919[Jackson] or Netty, many libraries store expensive objects in thread locals.
+These objects can only be accessed by the code running on the thread in which the objects are stored.
+Because the number of threads was limited, it capped the number of creation.
+Also, because threads were reused, the objects were cached and reused. 
+Unfortunately, these two assumptions are not valid with virtual threads:
+You can have a lot of them, they are not reused.
```

Comment:
even more: there's a clear indication from the Loom dev group which advertize against pooling them.